File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/palindrome-products/palindrome_products.go
package palindrome

import (
	"errors"
	"strconv"
)

type Product struct {
	Value          int
	Factorizations [][2]int
}

func Products(fmin, fmax int) (Product, Product, error) {
	if fmin > fmax {
		return Product{}, Product{}, errors.New("fmin > fmax")
	}

	productsMap := make(map[int][][2]int)

	for i := fmin; i <= fmax; i++ {
		for j := i; j <= fmax; j++ {
			product := i * j
			if product < 0 {
				continue
			}
			if isPalindrome(product) {
				productsMap[product] = append(productsMap[product], [2]int{i, j})
			}
		}
	}

	if len(productsMap) == 0 {
		return Product{}, Product{}, errors.New("no palindromes found")
	}

	var smallest, largest int
	first := true
	for p := range productsMap {
		if first {
			smallest, largest = p, p
			first = false
		} else {
			if p < smallest {
				smallest = p
			}
			if p > largest {
				largest = p
			}
		}
	}

	minProduct := Product{Value: smallest, Factorizations: productsMap[smallest]}
	maxProduct := Product{Value: largest, Factorizations: productsMap[largest]}
	return minProduct, maxProduct, nil
}

func isPalindrome(n int) bool {
	s := strconv.Itoa(n)
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/palindrome-products/palindrome-products.js
//
// This is only a SKELETON file for the 'Palindrome Products' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Palindromes {
  static generate({ minFactor, maxFactor }) {
    if (minFactor > maxFactor) {
      throw new Error('min must be <= max');
    }

    const palindromes = {};

    for (let a = minFactor; a <= maxFactor; a++) {
      for (let b = a; b <= maxFactor; b++) {
        const product = a * b;
        const str = product.toString();
        if (str === str.split('').reverse().join('')) {
          if (!palindromes[product]) {
            palindromes[product] = [];
          }
          palindromes[product].push([a, b]);
        }
      }
    }

    const products = Object.keys(palindromes).map(Number);
    if (products.length === 0) {
      return {
        smallest: { value: null, factors: [] },
        largest: { value: null, factors: [] }
      };
    }

    const minProduct = Math.min(...products);
    const maxProduct = Math.max(...products);

    return {
      smallest: { value: minProduct, factors: palindromes[minProduct] },
      largest: { value: maxProduct, factors: palindromes[maxProduct] },
    };
  }
}
